File: TUnit.TestProject/GenericMethodTests.cs
namespace TUnit.TestProject;

public class GenericMethodTests
{
    [Test]
    [MethodDataSource(nameof(AggregateBy_Numeric_TestData))]
    [MethodDataSource(nameof(AggregateBy_String_TestData))]
    public void AggregateBy_HasExpectedOutput<TSource, TKey, TAccumulate>(
        IEnumerable<TSource> source,
        Func<TSource, TKey> keySelector,
        Func<TKey, TAccumulate> seedSelector,
        Func<TAccumulate, TSource, TAccumulate> func,
        IEqualityComparer<TKey>? comparer,
        IEnumerable<KeyValuePair<TKey, TAccumulate>> expected) where TKey : notnull
    {
        var enumerable = source as TSource[] ?? source.ToArray();
        Console.WriteLine(string.Join(", ", enumerable));
        Console.WriteLine(string.Join(", ", enumerable.Select(keySelector)));
        Console.WriteLine(string.Join(", ", enumerable.Select(x => seedSelector(keySelector(x)))));
    }

    public static IEnumerable<(IEnumerable<int> source,
        Func<int, int> keySelector,
        Func<int, int> seedSelector,
        Func<int, int, int> func,
        IEqualityComparer<int>? comparer,
        IEnumerable<KeyValuePair<int, int>> expected)> AggregateBy_Numeric_TestData()
    {
        yield return (
            source: Enumerable.Range(0, 10),
            keySelector: x => x,
            seedSelector: x => 0,
            func: (x, y) => x + y,
            comparer: null,
            expected: Enumerable.Range(0, 10).ToDictionary(x => x, x => x)
        );
    }
    
    public static IEnumerable<(IEnumerable<string> source,
        Func<string, string> keySelector,
        Func<string, string> seedSelector,
        Func<string, string, string> func,
        IEqualityComparer<string>? comparer,
        IEnumerable<KeyValuePair<string, string>> expected)> AggregateBy_String_TestData()
    {
        yield return (
            source: ["Bob", "bob", "tim", "Bob", "Tim"],
            keySelector: x => x,
            seedSelector: x => "",
            func: (x, y) => x + y,
            comparer: null,
            expected: new Dictionary<string, string>(StringComparer.Ordinal)
            {
                { "Bob", "BobBob" },
                { "bob", "bob" },
                { "tim", "tim" },
                { "Tim", "Tim" },
            });
    }
}

File: TUnit.TestProject.FSharp/GenericMethodTests.fs
namespace TUnit.TestProject.FSharp

open System
open System.Collections.Generic
open TUnit.Core

// F# equivalent of GenericMethodTests.cs

type GenericMethodTests() =
    [<Test>]
    [<MethodDataSource("AggregateBy_Numeric_TestData")>]
    [<MethodDataSource("AggregateBy_String_TestData")>]
    member _.AggregateBy_HasExpectedOutput
        (source: seq<'TSource>,
         keySelector: 'TSource -> 'TKey,
         seedSelector: 'TKey -> 'TAccumulate,
         func: 'TAccumulate -> 'TSource -> 'TAccumulate,
         comparer: IEqualityComparer<'TKey>,
         expected: seq<KeyValuePair<'TKey, 'TAccumulate>>) =
        let enumerable = source |> Seq.toArray
        Console.WriteLine(String.Join(", ", enumerable))
        Console.WriteLine(String.Join(", ", enumerable |> Seq.map keySelector))
        Console.WriteLine(String.Join(", ", enumerable |> Seq.map (fun x -> seedSelector (keySelector x))))

    static member AggregateBy_Numeric_TestData() =
        seq {
            fun () ->
                let source = seq { 0 .. 9 }
                let keySelector = id
                let seedSelector = fun _ -> 0
                let func = fun x y -> x + y
                let comparer = null
                let expected =
                    seq { for x in 0 .. 9 -> KeyValuePair(x, x) }
                (source, keySelector, seedSelector, func, comparer, expected)
        }

    static member AggregateBy_String_TestData() =
        seq {
            fun () ->
                let source = [ "Bob"; "bob"; "tim"; "Bob"; "Tim" ]
                let keySelector = id
                let seedSelector = fun _ -> ""
                let func = fun x y -> x + y
                let comparer = null
                let expected =
                    [ KeyValuePair("Bob", "BobBob"); KeyValuePair("bob", "bob"); KeyValuePair("tim", "tim"); KeyValuePair("Tim", "Tim") ]
                (source, keySelector, seedSelector, func, comparer, expected)
        }
